feature: add `project_main_branch` resource

mdb opened the pull request:
~This is an experimental, work-in-progress PR to address issue #99 .~

~I've opened a draft PR to surface some visibility on my progress. I'm also interested in early feedback: am I on the right path, or is there more this feature would need to consider?~

Update: I believe this is ready for review, with the caveat that I don't believe I currently have access to the ability to [create a test-organization](https://github.com/rewe-digital/terraform-provider-sonarcloud#developing-the-provider) (Am I wrong?).

Thanks! :)

mdb commented on sonarcloud/resource_project_main_branch.go:
```diff
@@ -0,0 +1,203 @@
+	}
+
+	// NOTE: according to docs, this may not work for main branches, and may require admin privilege
+	// https://github.com/reinoudk/go-sonarcloud/blob/main/sonarcloud/project_branches/project_branches_gen.go#L5
```
☝️ Could/should `Delete` factor this into consideration and account for it in some way?

reinoudk replied:
Until #63 has been implemented it is not possible to run acceptance tests in CI, but if you create a test-organization as described [here](https://github.com/rewe-digital/terraform-provider-sonarcloud#developing-the-provider), you should be able to run the acceptance tests locally using `make testacc`.

As you put in the comment, `Delete` is only for non-main branches, so it cannot be used here. The NOOP would be possible (just remove it from state), but I'm wondering whether this resource should be a stand-alone resource. IMHO it would make more sense to have this as an attribute on the `sonarcloud_project` resource, i.e. `main_branch`.

mdb commented:
@reinoudk Before I invest more time, or potentially veer off-course, would you mind taking a preliminary look at this? Is this something you're amenable to? Am I offbase? Thanks!

reinoudk commented:
Hi @mdb, sorry for not spotting this one earlier. I'll have a look at it tonight.

reinoudk reviewed (commented):
Thanks for this PR! As I've said in the other comment, I think #99 is best solved by adding an attribute to the `sonarcloud_project` resource. 

Would you be to refactor this PR to add the `main_branch` attribute to the `sonarcloud_project` resource? You could reuse all the network code and keep the `findProjectMainBranch` function in `helpers.go`. Let me know what you think :)

reinoudk commented:
Yeah, I don't like using multiple endpoints for a single resource either. The thing is that the Sonarcloud API itself is not very REST-y and seems to mostly be there to service the WEB UI. For quite a few resources, there is no way to query for a single resource, so we instead have to list them all and then check if the one we want is in the response. That's pretty bonkers if you ask me 😅. The API just doesn't map that well to resources..

That being said, I would be OK with trying it out as a separate resource. More ergonomic usage could be achieved by writing a module that has the attribute and uses two resources underneath.

mdb commented:
> I would be OK with trying it out as a separate resource. More ergonomic usage could be achieved by writing a module that has the attribute and uses two resources underneath.

Thanks for the helpful response, @reinoudk . I _think_ I'm inclined to recommend the implementation remains as a distinct `project_main_branch`. Are you comfortable with that?

> Until https://github.com/rewe-digital/terraform-provider-sonarcloud/issues/63 has been implemented it is not possible to run acceptance tests in CI, but if you create a test-organization as described [here](https://github.com/rewe-digital/terraform-provider-sonarcloud#developing-the-provider), you should be able to run the acceptance tests locally using make testacc.

☝️ As for this, I don't _believe_ I have the ability to do so (am I wrong?). Is this something someone could do on my behalf? Asked a bit differently: what are next steps towards vetting/merging this PR? :)

reinoudk commented:
> Thanks for the helpful response, @reinoudk . I think I'm inclined to recommend the implementation remains as a distinct project_main_branch. Are you comfortable with that?

Yes, that sounds good!

>  As for this, I don't believe I have the ability to do so (am I wrong?).

You are right, this is something I have to find the time for to implement.

> Asked a bit differently: what are next steps towards vetting/merging this PR? :)

If you want to test the changes yourself, you can create a test organization as described in the README.md and use `make testacc` (the README also lists the env vars you need to set). I'll also try to build and test this branch this weekend.

reinoudk commented on sonarcloud/resource_project_main_branch.go:
```diff
@@ -0,0 +1,189 @@
+}
+
+func (r resourceProjectMainBranchType) NewResource(_ context.Context, p tfsdk.Provider) (tfsdk.Resource, diag.Diagnostics) {
+	return resourceProject{
```
Small copy-paste leftover ;)

```suggestion
	return resourceProjectMainBranch{
```

reinoudk commented on sonarcloud/resource_project_main_branch_test.go:
```diff
@@ -0,0 +1,58 @@
+resource "sonarcloud_project" "test" {
+	name = "%s"
+	key = "%s"
+	visibility = "visible"
```
Visibility should be either `public` or `private`.
```suggestion
	visibility = "public"
```

reinoudk replied:
I think this one's still in there 😄

reinoudk commented on sonarcloud/resource_project_main_branch.go:
```diff
@@ -0,0 +1,189 @@
+				Required:    true,
+				Description: "The name of the project main branch.",
+				PlanModifiers: tfsdk.AttributePlanModifiers{
+					tfsdk.RequiresReplace(),
```
Semantically I think it makes more sense to use `tfsdk.RequiresReplace()` on the `project_key`. Changing the name of the project main branch is an update, while changing the `project_key` would mean we actually start managing the attribute of a different resource. 

Practically there is not much of a difference, as the NOOP delete + create is the same as just an update.

reinoudk commented on sonarcloud/resource_project_main_branch.go:
```diff
@@ -0,0 +1,189 @@
+		return
+	}
+
+	projectKey := state.ProjectKey.Value
```
Nitpick: if changing the `project_key` causes a destroy (see my other comment regarding which attribute should semantically trigger a destroy), then it should never show up as changed in an update. It should be safe to remove this.

reinoudk reviewed (commented):
It's looking pretty good, thanks for all the effort so far on this! 🚀 I've added a few small comments, nothing big.

I've ran the test locally and I'm experiencing a race-condition with project-creation and setting the main branch of the project. This results in a 404 because the API is eventually consistent it seems. I'll have to see if I can fix this in the project resource with a retry mechanism for state refresh, but I'm a bit stressed for time at the moment.

If that should block this PR, we could change the test to use an existing project. I've set up an organization for #63 and need to wire it up in an action, but that should be done soon™ as well.

mdb commented:
> I'm experiencing a race-condition with project-creation and setting the main branch of the project.

@reinoudk If I understand you correctly, I believe I've corrected this such that `sonarcloud_project_main_branch.test` has an implicit dependency relationship on `sonarcloud_project.test` via `name = sonarcloud_project.test.name`. Would that address the race condition?

(And I believe I've addressed your other comments, too - thanks! :) )

reinoudk commented on sonarcloud/resource_project_main_branch_test.go:
```diff
@@ -0,0 +1,58 @@
+}
+
+resource "sonarcloud_project_main_branch" "test" {
+	name = sonarcloud_project.test.name
```
Ah, here the name is not the name of the project, but the name of the branch. I didn't see that the other time either.
So this would be 
```
name = "%s%
project_key = sonarcloud_project.test.key
```

(Note that it should be `.key` and not `name`, I always mix those up as well 😅).

mdb replied:
Ah, good call! :)

reinoudk commented:
> @reinoudk If I understand you correctly, I believe I've corrected this such that `sonarcloud_project_main_branch.test` has an implicit dependency relationship on `sonarcloud_project.test` via `name = sonarcloud_project.test.name`. Would that address the race condition?

I think it is a race condition in the API itself, but I have to verify this. This issue is still there, and even with a small fix to the `project_key` reference (see other comment), this was throwing a 404.

mdb commented:
@reinoudk Thanks for the great reviews and apologies for the careless errors on my end. I believe this is ready for a re-review!

reinoudk reviewed (approved):
> @reinoudk Thanks for the great reviews and apologies for the careless errors on my end. I believe this is ready for a re-review!

No worries, these small things happen to me all the time 😄 . The code looks good but I'm still having some issues with the test. 

So I'll approve and merge this, but I want to fix the race condition before I'm creating the `0.4.0` release. I should be able to do that this weekend, sorry for the wait.
